### PR TITLE
Added payment form to checkout page

### PR DESCRIPTION
**Pull request by clintonb**
Learners can now input credit card details directly on the checkout page instead of redirecting to CyberSource.

SOL-2117

### CODE REVIEW

**Comment by clintonb** [comment]
Depends on #983

**Comment by clintonb** [comment]
@mjfrey @iivic @marjev @vkaracic  please review
FYI @edx/ecommerce

File: ecommerce/extensions/basket/views.py [hunk 1]
```diff
@@ -142,14 +147,23 @@ def get_context_data(self, **kwargs):
...
                 'analytics_data': prepare_analytics_data(
-                    self.request.user,
+                    user,
                     self.request.site.siteconfiguration.segment_key,
                     unicode(course_key)
                 ),
+                'enable_client_side_checkout': False,
             })
 
+            if waffle.flag_is_active(self.request, 'enable_client_side_checkout'):
```
Comment (mjfrey) [inline ecommerce/extensions/basket/views.py]: I think we want to be able to control client side checkout per site.  This should be in the site_configuration.  I also think if this is enabled for a site, a sample should control the rollout.

File: ecommerce/extensions/basket/views.py [hunk 2]
```diff
@@ -142,14 +147,23 @@ def get_context_data(self, **kwargs):
...
+                context.update({
+                    'enable_client_side_checkout': True,
+                    'payment_form': PaymentForm(user=user, initial={'basket': basket}, label_suffix=''),
+                    'payment_url': CybersourceSOP().sop_url,
```
Comment (mjfrey) [inline ecommerce/extensions/basket/views.py]: Can we make this generic somehow?  What if we have more than 1 payment provider that supports client side.

File: ecommerce/static/js/apps/basket_app.js [hunk 3]
```diff
@@ -6,6 +6,86 @@ require([
               BasketPage) {
         'use strict';
 
-        $(document).ready(BasketPage.onReady);
+        // TODO Remove this before merging!
```
Comment (vkaracic) [inline ecommerce/static/js/apps/basket_app.js]: Can this be removed now?

File: ecommerce/static/js/apps/basket_app.js [hunk 4]
```diff
@@ -6,6 +6,86 @@ require([
...
+                        for (var key in formData) {
+                            if (formData.hasOwnProperty(key)) {
+                                $paymentForm.append(
+                                    '<input type="hidden" name="' + key + '" value="' + formData[key] + '" />'
+                                );
+                            }
+                        }
+                    },
+                    error: function (jqXHR, textStatus, errorThrown) {
+                        // TODO Handle errors. Ideally the form should be validated in JavaScript
```
Comment (vkaracic) [inline ecommerce/static/js/apps/basket_app.js:48]: When is this going to be done?

File: ecommerce/static/js/apps/basket_app.js [hunk 5]
```diff
@@ -6,6 +6,86 @@ require([
...
+
+                // Post synchronously since we need the returned data.
+                $.ajax({
+                    type: 'POST',
+                    url: signingUrl,
+                    data: $signedFields.serialize(),
+                    async: false,
+                    success: function (data, textStatus, jqXHR) {
+                        var formData = data.form_fields;
+                        console.log(data);
```
Comment (vkaracic) [inline ecommerce/static/js/apps/basket_app.js]: Are the logs here and below still needed?

**Review by vkaracic — COMMENTED** [review]
Since it's a change in the UI, can we see some screenshots?
Never mind, I understand that the UI will be an additional ticket.

File: ecommerce/extensions/basket/views.py (hunk 2 above)
Comment (clintonb) [reply ecommerce/extensions/basket/views.py]: This is a prototype. Making it more generic is premature. Additionally, I can't envision a scenario in which we would have two distinct credit card processors.

File: ecommerce/static/js/apps/basket_app.js (hunk 3 above)
Comment (clintonb) [reply ecommerce/static/js/apps/basket_app.js]: Done.

File: ecommerce/static/js/apps/basket_app.js (hunk 5 above)
Comment (clintonb) [reply ecommerce/static/js/apps/basket_app.js]: Removed.

File: ecommerce/static/js/apps/basket_app.js (hunk 4 above)
Comment (clintonb) [reply ecommerce/static/js/apps/basket_app.js:48]: Whenever the front-end work is completed. This is simply a prototype.

**Comment by clintonb** [comment]
@mjfrey @iivic @marjev @vkaracic any other feedback?

File: ecommerce/templates/oscar/checkout/thank_you.html [hunk 6]
```diff
@@ -0,0 +1,328 @@
+{# The layout here is pretty much the same as Oscar. The following changes have been made:
```
Comment (mjfrey) [inline ecommerce/templates/oscar/checkout/thank_you.html:1]: will this be replaced with our new receipt page?

File: ecommerce/templates/oscar/basket/partials/basket_content.html [hunk 7]
```diff
@@ -151,6 +152,68 @@
...
+                                                {% trans "CVV" %}
+                                            </label>
+                                            <input id="id_card_cvn" name="card_cvn" class="form-control pci-field"
+                                                   maxlength="4"/>
+                                        </div>
+                                    </div>
+
+                                    <button type="submit" class="btn btn-primary btn-large" data-track-type="click"
+                                            data-track-event="edx.bi.ecommerce.basket.payment_selected"
+                                            data-track-category="checkout" data-processor-name="cybersource_sop"
```
Comment (mjfrey) [inline ecommerce/templates/oscar/basket/partials/basket_content.html:208]: can this be data-processor-name=processor.NAME + 'client'?

File: ecommerce/templates/oscar/checkout/thank_you.html (hunk 6 above)
Comment (clintonb) [reply ecommerce/templates/oscar/checkout/thank_you.html:1]: I don't know what template was used. Given the lack of conflicts, I'm guessing not. That being said, this template is mostly here for my own edification.

File: ecommerce/templates/oscar/basket/partials/basket_content.html (hunk 7 above)
Comment (clintonb) [reply ecommerce/templates/oscar/basket/partials/basket_content.html:208]: We are going to have to hardcode CyberSource info either way. See the usage of `cybersource_submit`. This is where I pull the prototype card and assure you we'll clean this up later when we have a better idea of how plugins might work.

**Review by iivic — APPROVED** [review]

**Review by vkaracic — APPROVED** [review]

**Review by marjev — APPROVED** [review]

**Review by mjfrey — APPROVED** [review]

**Pull request merged**